File: go/persistenttrafficlight/trafficlight/mom.go
package trafficlight

import (
	"github.com/frame-lang/frame-demos/persistenttrafficlight/framelang"
)

func NewTrafficLightMom() TrafficLightMom {
	m := &trafficLightMomStruct{}

	// Validate interfaces
	var _ TrafficLightMom = m
	var _ TrafficLightMom_actions = m

	// Create and intialize start state compartment.
	m._compartment_ = NewTrafficLightMomCompartment(TrafficLightMomState_New)

	// Override domain variables.
	m.trafficLight = nil
	m.data = nil

	// Send system start event
	e := framelang.FrameEvent{Msg: ">"}
	m._mux_(&e)
	return m
}

type TrafficLightMomState uint

const (
	TrafficLightMomState_New TrafficLightMomState = iota
	TrafficLightMomState_Saving
	TrafficLightMomState_Persisted
	TrafficLightMomState_Working
	TrafficLightMomState_TrafficLightApi
	TrafficLightMomState_End
)

type TrafficLightMom interface {
	Stop()
	Tick()
	EnterRed()
	EnterGreen()
	EnterYellow()
	EnterFlashingRed()
	ExitFlashingRed()
	StartWorkingTimer()
	StopWorkingTimer()
	StartFlashingTimer()
	StopFlashingTimer()
	ChangeColor(color string)
	StartFlashing()
	StopFlashing()
	ChangeFlashingAnimation()
	SystemError()
	SystemRestart()
	Log(msg string)
}

type TrafficLightMom_actions interface {
	enterRed()
	enterGreen()
	enterYellow()
	enterFlashingRed()
	exitFlashingRed()
	startWorkingTimer()
	stopWorkingTimer()
	startFlashingTimer()
	stopFlashingTimer()
	changeColor(color string)
	startFlashing()
	stopFlashing()
	changeFlashingAnimation()
	systemError()
	systemRestart()
	log(msg string)
}

type trafficLightMomStruct struct {
	_compartment_     *TrafficLightMomCompartment
	_nextCompartment_ *TrafficLightMomCompartment
	trafficLight      TrafficLight
	data              []byte
}

//===================== Interface Block ===================//

func (m *trafficLightMomStruct) Stop() {
	e := framelang.FrameEvent{Msg: "stop"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) Tick() {
	e := framelang.FrameEvent{Msg: "tick"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) EnterRed() {
	e := framelang.FrameEvent{Msg: "enterRed"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) EnterGreen() {
	e := framelang.FrameEvent{Msg: "enterGreen"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) EnterYellow() {
	e := framelang.FrameEvent{Msg: "enterYellow"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) EnterFlashingRed() {
	e := framelang.FrameEvent{Msg: "enterFlashingRed"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) ExitFlashingRed() {
	e := framelang.FrameEvent{Msg: "exitFlashingRed"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) StartWorkingTimer() {
	e := framelang.FrameEvent{Msg: "startWorkingTimer"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) StopWorkingTimer() {
	e := framelang.FrameEvent{Msg: "stopWorkingTimer"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) StartFlashingTimer() {
	e := framelang.FrameEvent{Msg: "startFlashingTimer"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) StopFlashingTimer() {
	e := framelang.FrameEvent{Msg: "stopFlashingTimer"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) ChangeColor(color string) {
	params := make(map[string]interface{})
	params["color"] = color
	e := framelang.FrameEvent{Msg: "changeColor", Params: params}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) StartFlashing() {
	e := framelang.FrameEvent{Msg: "startFlashing"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) StopFlashing() {
	e := framelang.FrameEvent{Msg: "stopFlashing"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) ChangeFlashingAnimation() {
	e := framelang.FrameEvent{Msg: "changeFlashingAnimation"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) SystemError() {
	e := framelang.FrameEvent{Msg: "systemError"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) SystemRestart() {
	e := framelang.FrameEvent{Msg: "systemRestart"}
	m._mux_(&e)
}

func (m *trafficLightMomStruct) Log(msg string) {
	params := make(map[string]interface{})
	params["msg"] = msg
	e := framelang.FrameEvent{Msg: "log", Params: params}
	m._mux_(&e)
}

//====================== Multiplexer ====================//

func (m *trafficLightMomStruct) _mux_(e *framelang.FrameEvent) {
	switch m._compartment_.State {
	case TrafficLightMomState_New:
		m._TrafficLightMomState_New_(e)
	case TrafficLightMomState_Saving:
		m._TrafficLightMomState_Saving_(e)
	case TrafficLightMomState_Persisted:
		m._TrafficLightMomState_Persisted_(e)
	case TrafficLightMomState_Working:
		m._TrafficLightMomState_Working_(e)
	case TrafficLightMomState_TrafficLightApi:
		m._TrafficLightMomState_TrafficLightApi_(e)
	case TrafficLightMomState_End:
		m._TrafficLightMomState_End_(e)
	}

	if m._nextCompartment_ != nil {
		nextCompartment := m._nextCompartment_
		m._nextCompartment_ = nil
		if nextCompartment._forwardEvent_ != nil &&
			nextCompartment._forwardEvent_.Msg == ">" {
			m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
			m._compartment_ = nextCompartment
			m._mux_(nextCompartment._forwardEvent_)
		} else {
			m._do_transition_(nextCompartment)
			if nextCompartment._forwardEvent_ != nil {
				m._mux_(nextCompartment._forwardEvent_)
			}
		}
		nextCompartment._forwardEvent_ = nil
	}
}

//===================== Machine Block ===================//

func (m *trafficLightMomStruct) _TrafficLightMomState_New_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		m.trafficLight = NewTrafficLight(m)
		// Traffic Light\nStarted
		compartment := NewTrafficLightMomCompartment(TrafficLightMomState_Saving)
		m._transition_(compartment)
		return
	}
	m._TrafficLightMomState_TrafficLightApi_(e)

}

func (m *trafficLightMomStruct) _TrafficLightMomState_Saving_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		m.data = m.trafficLight.Marshal()
		m.trafficLight = nil
		// Saved
		compartment := NewTrafficLightMomCompartment(TrafficLightMomState_Persisted)
		m._transition_(compartment)
		return
	}
}

func (m *trafficLightMomStruct) _TrafficLightMomState_Persisted_(e *framelang.FrameEvent) {
	switch e.Msg {
	case "tick":
		// Tick
		compartment := NewTrafficLightMomCompartment(TrafficLightMomState_Working)
		compartment._forwardEvent_ = e
		m._transition_(compartment)
		return
	case "systemError":
		// System Error
		compartment := NewTrafficLightMomCompartment(TrafficLightMomState_Working)
		compartment._forwardEvent_ = e
		m._transition_(compartment)
		return
	case "stop":
		// Stop
		compartment := NewTrafficLightMomCompartment(TrafficLightMomState_End)
		m._transition_(compartment)
		return
	}
}

func (m *trafficLightMomStruct) _TrafficLightMomState_Working_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		m.trafficLight = LoadTrafficLight(m, m.data)
		return
	case "tick":
		m.trafficLight.Tick()
		// Done
		compartment := NewTrafficLightMomCompartment(TrafficLightMomState_Saving)
		m._transition_(compartment)
		return
	case "systemError":
		m.trafficLight.SystemError()
		// Done
		compartment := NewTrafficLightMomCompartment(TrafficLightMomState_Saving)
		m._transition_(compartment)
		return
	}
	m._TrafficLightMomState_TrafficLightApi_(e)

}

func (m *trafficLightMomStruct) _TrafficLightMomState_TrafficLightApi_(e *framelang.FrameEvent) {
	switch e.Msg {
	case "enterRed":
		m.enterRed()
		return
	case "enterGreen":
		m.enterGreen()
		return
	case "enterYellow":
		m.enterYellow()
		return
	case "enterFlashingRed":
		m.enterFlashingRed()
		return
	case "exitFlashingRed":
		m.exitFlashingRed()
		return
	case "startWorkingTimer":
		m.startWorkingTimer()
		return
	case "stopWorkingTimer":
		m.stopWorkingTimer()
		return
	case "startFlashingTimer":
		m.startFlashingTimer()
		return
	case "stopFlashingTimer":
		m.stopFlashingTimer()
		return
	case "changeColor":
		m.changeColor(e.Params["color"].(string))
		return
	case "startFlashing":
		m.startFlashing()
		return
	case "stopFlashing":
		m.stopFlashing()
		return
	case "changeFlashingAnimation":
		m.changeFlashingAnimation()
		return
	case "systemError":
		m.systemError()
		return
	case "systemRestart":
		m.systemRestart()
		return
	case "log":
		m.log(e.Params["msg"].(string))
		return
	}
}

func (m *trafficLightMomStruct) _TrafficLightMomState_End_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		m.trafficLight = LoadTrafficLight(m, m.data)
		m.trafficLight.Stop()
		m.trafficLight = nil
		return
	}
	m._TrafficLightMomState_TrafficLightApi_(e)

}

//=============== Machinery and Mechanisms ==============//

func (m *trafficLightMomStruct) _transition_(compartment *TrafficLightMomCompartment) {
	m._nextCompartment_ = compartment
}

func (m *trafficLightMomStruct) _do_transition_(nextCompartment *TrafficLightMomCompartment) {
	m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
	m._compartment_ = nextCompartment
	m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}

//===================== Actions Block ===================//

/********************************************************

// Unimplemented Actions

func (m *trafficLightMomStruct) enterRed()  {}
func (m *trafficLightMomStruct) enterGreen()  {}
func (m *trafficLightMomStruct) enterYellow()  {}
func (m *trafficLightMomStruct) enterFlashingRed()  {}
func (m *trafficLightMomStruct) exitFlashingRed()  {}
func (m *trafficLightMomStruct) startWorkingTimer()  {}
func (m *trafficLightMomStruct) stopWorkingTimer()  {}
func (m *trafficLightMomStruct) startFlashingTimer()  {}
func (m *trafficLightMomStruct) stopFlashingTimer()  {}
func (m *trafficLightMomStruct) changeColor(color string)  {}
func (m *trafficLightMomStruct) startFlashing()  {}
func (m *trafficLightMomStruct) stopFlashing()  {}
func (m *trafficLightMomStruct) changeFlashingAnimation()  {}
func (m *trafficLightMomStruct) systemError()  {}
func (m *trafficLightMomStruct) systemRestart()  {}
func (m *trafficLightMomStruct) log(msg string)  {}

********************************************************/

//=============== Compartment ==============//

type TrafficLightMomCompartment struct {
	State          TrafficLightMomState
	StateArgs      map[string]interface{}
	StateVars      map[string]interface{}
	EnterArgs      map[string]interface{}
	ExitArgs       map[string]interface{}
	_forwardEvent_ *framelang.FrameEvent
}

func NewTrafficLightMomCompartment(state TrafficLightMomState) *TrafficLightMomCompartment {
	c := &TrafficLightMomCompartment{State: state}
	c.StateArgs = make(map[string]interface{})
	c.StateVars = make(map[string]interface{})
	c.EnterArgs = make(map[string]interface{})
	c.ExitArgs = make(map[string]interface{})
	return c
}

File: javascript/persistenttrafficlight/trafficlight/mom.js
// emitted from framec_v0.10.0

const TrafficLightController = require("./machine");
const FrameEvent = require("../framelang/framelang");
// get include files at https://github.com/frame-lang/frame-ancillary-files

class TrafficLightMom {
    
    constructor () {
        
        // Create and intialize start state compartment.
        
        this._state = this._sNew_;
        this._compartment = new TrafficLightMomCompartment(this._state);
        this._nextCompartment = null;
        
        // Initialize domain
        this.trafficLight = null;
        this.data = null;
        
        // Send system start event
        this._frameEvent = FrameEvent(">", null);
        this._mux_(this._frameEvent);
    }
    
    //===================== Interface Block ===================//
    
    stop() {
        let e = FrameEvent("stop",null);
        this._mux_(e);
    }
    
    tick() {
        let e = FrameEvent("tick",null);
        this._mux_(e);
    }
    
    enterRed() {
        let e = FrameEvent("enterRed",null);
        this._mux_(e);
    }
    
    enterGreen() {
        let e = FrameEvent("enterGreen",null);
        this._mux_(e);
    }
    
    enterYellow() {
        let e = FrameEvent("enterYellow",null);
        this._mux_(e);
    }
    
    enterFlashingRed() {
        let e = FrameEvent("enterFlashingRed",null);
        this._mux_(e);
    }
    
    exitFlashingRed() {
        let e = FrameEvent("exitFlashingRed",null);
        this._mux_(e);
    }
    
    startWorkingTimer() {
        let e = FrameEvent("startWorkingTimer",null);
        this._mux_(e);
    }
    
    stopWorkingTimer() {
        let e = FrameEvent("stopWorkingTimer",null);
        this._mux_(e);
    }
    
    startFlashingTimer() {
        let e = FrameEvent("startFlashingTimer",null);
        this._mux_(e);
    }
    
    stopFlashingTimer() {
        let e = FrameEvent("stopFlashingTimer",null);
        this._mux_(e);
    }
    
    changeColor(color) {
        let e = FrameEvent("changeColor",{"color":color});
        this._mux_(e);
    }
    
    startFlashing() {
        let e = FrameEvent("startFlashing",null);
        this._mux_(e);
    }
    
    stopFlashing() {
        let e = FrameEvent("stopFlashing",null);
        this._mux_(e);
    }
    
    changeFlashingAnimation() {
        let e = FrameEvent("changeFlashingAnimation",null);
        this._mux_(e);
    }
    
    systemError() {
        let e = FrameEvent("systemError",null);
        this._mux_(e);
    }
    
    systemRestart() {
        let e = FrameEvent("systemRestart",null);
        this._mux_(e);
    }
    
    log(msg) {
        let e = FrameEvent("log",{"msg":msg});
        this._mux_(e);
    }
    
    //====================== Multiplexer ====================//
    
    _mux_(e) {
        switch (this._compartment.state) {
            case this._sNew_:
                this._sNew_(e);
                break;
            case this._sSaving_:
                this._sSaving_(e);
                break;
            case this._sPersisted_:
                this._sPersisted_(e);
                break;
            case this._sWorking_:
                this._sWorking_(e);
                break;
            case this._sTrafficLightApi_:
                this._sTrafficLightApi_(e);
                break;
            case this._sEnd_:
                this._sEnd_(e);
                break;
        }
        
        if( this._nextCompartment != null) {
            let nextCompartment = this._nextCompartment
            this._nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this._mux_(FrameEvent( "<", this._compartment.ExitArgs))
                this._compartment = nextCompartment
                this._mux_(nextCompartment._forwardEvent)
            } else {
                this._do_transition_(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this._mux_(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    _sNew_(e) {
        switch (e._message) {
            case ">":
                {
                this.trafficLight = new TrafficLightController(this);
                // Traffic Light\nStarted
                let compartment =  new TrafficLightMomCompartment(this._sSaving_);
                
                
                this._transition_(compartment);
                return;
                }
                
        }
        this._sTrafficLightApi_(e);
        
    }
    
    _sSaving_(e) {
        switch (e._message) {
            case ">":
                {
                this.data = this.trafficLight.marshal();
                this.trafficLight = null;
                // Saved
                let compartment =  new TrafficLightMomCompartment(this._sPersisted_);
                
                
                this._transition_(compartment);
                return;
                }
                
        }
    }
    
    _sPersisted_(e) {
        switch (e._message) {
            case "tick":
                {
                // Tick
                let compartment =  new TrafficLightMomCompartment(this._sWorking_);
                
                compartment._forwardEvent = e;
                
                this._transition_(compartment);
                return;
                }
                
            case "systemError":
                {
                // System Error
                let compartment =  new TrafficLightMomCompartment(this._sWorking_);
                
                compartment._forwardEvent = e;
                
                this._transition_(compartment);
                return;
                }
                
            case "stop":
                {
                // Stop
                let compartment =  new TrafficLightMomCompartment(this._sEnd_);
                
                
                this._transition_(compartment);
                return;
                }
                
        }
    }
    
    _sWorking_(e) {
        switch (e._message) {
            case ">":
                {
                this.trafficLight = TrafficLightController.loadTrafficLight(this,this.data);
                return;
                }
                
            case "tick":
                {
                this.trafficLight.tick();
                // Done
                let compartment =  new TrafficLightMomCompartment(this._sSaving_);
                
                
                this._transition_(compartment);
                return;
                }
                
            case "systemError":
                {
                this.trafficLight.systemError_do();
                // Done
                let compartment =  new TrafficLightMomCompartment(this._sSaving_);
                
                
                this._transition_(compartment);
                return;
                }
                
        }
        this._sTrafficLightApi_(e);
        
    }
    
    _sTrafficLightApi_(e) {
        switch (e._message) {
            case "enterRed":
                {
                this.enterRed_do();
                return;
                }
                
            case "enterGreen":
                {
                this.enterGreen_do();
                return;
                }
                
            case "enterYellow":
                {
                this.enterYellow_do();
                return;
                }
                
            case "enterFlashingRed":
                {
                this.enterFlashingRed_do();
                return;
                }
                
            case "exitFlashingRed":
                {
                this.exitFlashingRed_do();
                return;
                }
                
            case "startWorkingTimer":
                {
                this.startWorkingTimer_do();
                return;
                }
                
            case "stopWorkingTimer":
                {
                this.stopWorkingTimer_do();
                return;
                }
                
            case "startFlashingTimer":
                {
                this.startFlashingTimer_do();
                return;
                }
                
            case "stopFlashingTimer":
                {
                this.stopFlashingTimer_do();
                return;
                }
                
            case "changeColor":
                {
                this.changeColor_do((e._parameters["color"]));
                return;
                }
                
            case "startFlashing":
                {
                this.startFlashing_do();
                return;
                }
                
            case "stopFlashing":
                {
                this.stopFlashing_do();
                return;
                }
                
            case "changeFlashingAnimation":
                {
                this.changeFlashingAnimation_do();
                return;
                }
                
            case "systemError":
                {
                this.systemError_do();
                return;
                }
                
            case "systemRestart":
                {
                this.systemRestart_do();
                return;
                }
                
            case "log":
                {
                this.log_do((e._parameters["msg"]));
                return;
                }
                
        }
    }
    
    _sEnd_(e) {
        switch (e._message) {
            case ">":
                {
                this.trafficLight = TrafficLightController.loadTrafficLight(this,this.data);
                this.trafficLight.stop();
                this.trafficLight = null;
                return;
                }
                
        }
        this._sTrafficLightApi_(e);
        
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    enterRed_do() { throw new Error('Action not implemented.'); }
    enterGreen_do() { throw new Error('Action not implemented.'); }
    enterYellow_do() { throw new Error('Action not implemented.'); }
    enterFlashingRed_do() { throw new Error('Action not implemented.'); }
    exitFlashingRed_do() { throw new Error('Action not implemented.'); }
    startWorkingTimer_do() { throw new Error('Action not implemented.'); }
    stopWorkingTimer_do() { throw new Error('Action not implemented.'); }
    startFlashingTimer_do() { throw new Error('Action not implemented.'); }
    stopFlashingTimer_do() { throw new Error('Action not implemented.'); }
    changeColor_do(color) { throw new Error('Action not implemented.'); }
    startFlashing_do() { throw new Error('Action not implemented.'); }
    stopFlashing_do() { throw new Error('Action not implemented.'); }
    changeFlashingAnimation_do() { throw new Error('Action not implemented.'); }
    systemError_do() { throw new Error('Action not implemented.'); }
    systemRestart_do() { throw new Error('Action not implemented.'); }
    log_do(msg) { throw new Error('Action not implemented.'); }
    
    //=============== Machinery and Mechanisms ==============//
    
    _transition_(compartment) {
        this._nextCompartment = compartment;
    }
    
    _do_transition_(nextCompartment) {
        this._mux_(FrameEvent("<", this._compartment.ExitArgs));
        this._compartment = nextCompartment;
        this._mux_(FrameEvent(">", this._compartment.EnterArgs));
    }
    
    
    
};


//=============== Compartment ==============//

class TrafficLightMomCompartment {

    constructor(state) {
        this.state = state
    }
    
    StateArgs = {};
    StateVars = {};
    EnterArgs = {};
    ExitArgs = {};
    _forwardEvent = FrameEvent.call(this)
}


// /********************

class TrafficLightMomController extends TrafficLightMom {

	constructor() {
	  super()
	}

	enterRed_do() {
        console.log("Red")
    }
	enterGreen_do() {
        console.log("Green")
    }
	enterYellow_do() {
        console.log("Yellow")
    }
	enterFlashingRed_do() {}
	exitFlashingRed_do() {}
	startWorkingTimer_do() {}
	stopWorkingTimer_do() {}
	startFlashingTimer_do() {}
	stopFlashingTimer_do() {}
	changeColor_do(color) {}
	startFlashing_do() {}
	stopFlashing_do() {}
	changeFlashingAnimation_do() {}
	systemError_do() {}
	systemRestart_do() {}
	log_do(msg) {}

};

// ********************/


module.exports = TrafficLightMomController